modules/caddyhttp/reverseproxy/httptransport.go

```diff
@@ -166,6 +165,9 @@ func (h *HTTPTransport) setScheme(req *http.Request) {
 
 // Cleanup implements caddy.CleanerUpper and closes any idle connections.
 func (h HTTPTransport) Cleanup() error {
+	if h.Transport == nil {
+		return nil
+	}
```

Comment:
This prevents a panic in the case where `Provision` fails and `h.Transport` is left `nil` during the `Cleanup()` call 
https://github.com/caddyserver/caddy/blob/v2/context.go#L327

Reply:
Shouldn't this be an error since this should never happen? I get the desire to reduce the possibility of panic but if Provisioning fails, is the server working as intended?

Reply:
I can't speak authoritatively, suffice to say that in the context of cleanup I wouldn't consider this an error but rather the absence of anything to clean up.

Reply:
This is defensive programming to allow a potential bug to go un-noticed, IMHO. This is the kind of code that scares me. If `h.Transport` should never be `nil`, or if it is `nil` this call should not be made, then the application should log these scenarios. Returning an error is important.

Reply:
Seems fair, I couldn't say if this should never happen or not, but given your confidence and the fact that it's the `Transport` field on a struct called `HTTPTransport` that stance seems reasonable. 

Thoughts on:
```
    return errors.New("transport is nil, nothing to cleanup")
```

For context: from what I recall the code that calls this is largely ignorant to the modules themselves. If `Provision` fails, it simply runs the `Cleanup` logic as if it was shutting down. I'm not readily familiar enough to reference the relevant blocks offhand, but I'm not quite sure what the config loading logic should do differently, in the event that `Provision` fails (like due to a bad config, which is how I hit this), how would you know whether or not there's something to cleanup given the abstractions at play? Provision could return more information perhaps, or even a special error indicating cleanup needs, but that seems overly complex. Just fruit for thought ❤️ I know you like it

Reply:
Ah, good catch!

Great discussion here; all good points. You're both right. :)

Normally, I avoid (overly) defensive programming as well, since errors should be exposed in the proper place. Indeed, this is not the right place to raise an error.

However, this function _does_ need to be defensive.

To clarify, `Cleanup()` **is supposed** to be defensive since it is called even if `Provision()` returns an error. From the [godoc](https://pkg.go.dev/github.com/caddyserver/caddy/v2@v2.0.0-beta11?tab=doc#CleanerUpper):

> Cleanup should work even if Provision returns an error, to allow cleaning up from partial provisionings.

The reason this is necessary is because `Provision()` may partially complete before returning an error; i.e. there's no way to guarantee that an error'ed `Provision()` call has rolled itself back. And duplicating that logic in a defer in `Provision()` and also in `Cleanup()` would be undesirable, tedious, and error-prone. The easiest thing to do I think is to simply have a defensive `Cleanup()` instead.

In short: this nil check is good, but we shouldn't return an error if `Cleanup` is nil. In that case, there is simply nothing to clean up and it may be a no-op. If there was an error earlier that prevented the Transport from being set, it should have already been reported as an error, hence (possibly) why this function is being called.

Hope that makes sense! Remove the error return here and this will be good.